README
******************************************************************
*                   http://software.pt-link.net/                 *
******************************************************************
* $Id: README,v 1.3 2005/08/27 16:23:49 jpinto Exp $           *
Install instructions: 
  INSTALL
License details:
  LICENSE
Changes Log:
  CHANGES
Inter-server protocol details:
  doc/PTS4.txt
helpsys data files:
  doc/help.*
  NOTE:
  helpsys entries must start with '*'item
  a single '*' represents the default helpsys output (no item)

HOW TO GET HELP:
----------------
 [ http://www.ptlink.net/Coders/ ]
                                                                                
1.  Check the file ircd.log
  ( Learn english if you don't understand the problem :P )
                                                                                
2. Post your question on PTlink Coders forum on the web page.
                                                                                
3. The Mailing List
There is a mailing list for general discussion of PTlink IRCd,
you can subscribe it from http://www.ptlink.net/Coders/
If PTlink6 crashes on you, PLEASE use the ./sendbug script.
The PTlink Coders Team can't fix bugs if no one tells about them!
                                                                                
4. The support/suggestions channel
Connect to irc.ptlink.net:6667 and join #PTC (english only).
                                                                                
If there is a new feature you would like to see in the next major
release of PTlink IRCd, please post it to the public mailing list,
ptlinksoft-ircd@lists.sourceforge.net


CHANGES
******************************************************************
*                   http://software.pt-link.net                  *
******************************************************************

(.6) 08 May 2006

  What is new ?
  ---------------------
  #79, option to request login code for human verification (/HVC ON)
  #85, missing NoSpamExitMsg option will crash the ircd

(.5) 07 Jan 2006

  What is new ?
  ---------------------
  #57: DefaultChanMode option to define default channel modes
  #55: super vlines should be checked on away messages
  #32:  configure --enable-halfops option to enable half-ops
  0000346: WebPass option to be used by webchat clients  
  Add rehash script to the tools dir

  What has been fixed ?
  ---------------------
  #13 : chanserv identify is now removed from /identify
  #51 : Compiling now works on x64 systems
  #6 : /msg botserv displays garbage  
  0000347: supersvlines should use match patterns
  0000324: PIDFILE location is now used when specific location is given


(.4) 27 Aug 2005

  What is new ?
  ---------------------
  0000313: configure option to enable/disable flood delay
  0000301: super svlines
  0000263: chghost command for remote hostname change
  0000260: utf-8 support on nicks and chan names
  Added reconf.sh script to call autoconf
  GLineOnSVline: new dconf option to automaticly gline who trigger a svline

  What has been fixed ?
  ---------------------
  0000292: crash if removing +o from users without o:line
  0000289: crash on SVSJOIN/SVSPART with missing parameter
  0000286: gline #chan crash on a rare condition
  0000264: compilation with gcc4 fails

(.3) 28 Jan 2005

  What is new ?
  ---------------------
  0000184: CryptKey dconf option replaced services timestamp for spoofing
  0000186: Add /ns, /cs, /os etc. as services aliases

  What has been fixed ?
  ---------------------
  0000189: Incorrect message source on debug option (-d) logs
  0000182: -d was adedd to the invalid syntax message
  0000190: If CryptKey is disabled keep using old situation of randomness

(.2) 11 Dec 2004

  What has been fixed ?
  ---------------------
  0000181: remove unused SCS code
  0000171: remove the auto-away when the idle time is over
  0000177: Do not allow "\" to be used on nicks, chans and usernames

(.1) 13 Nov 2004

  What is new ?
  ---------------------
  0000162: domain stats
  0000167: MaxOperIdleTime
  0000168: AutoAwayIdleTime

  What has been fixed ?
  ---------------------
  0000170: StrongVlines option should be removed
  0000164: AutoJoinChan with multiple chans only works for first user
  0000166: strip colors instead of blocking them on chanmode +c
  0000163: PTS4.txt out of date

  PTlink6.19.0 (05 Nov 2004)
==================================================================

  What is new ?
  ---------------------
  Issue 0000151: AutoJoinChan dconf option
  Issue 0000144: MaxChansPerRegUser to specify maxchans for registered users
  Issue 0000130: Protocol change to use SNICK/SSJOIN on netjoins
  Issue 0000143: killircd tool to be used on windows version
  Issue 0000128: command line debug option

  What has been fixed ?
  ---------------------
  Issue 0000120: warning compiling with ipv6 support
  Issue 0000133: SVLINES should filter words on AWAY reasons
  Issue 0000135: GLineOnExcessFlood is a timed option
  Issue 0000136: dconf now use macros
  Issue 0000091: fixed again (old fix wasn't even compiling) 

  To Do
  -----
  Merge TKlines code from Hybrid6.3
  Fix RestrictedChans 
	(empty channels are not synchronized on netjoins)


src/md5.c
/******** Code specific to the type of encryption. ********/
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
 
/* Copyright (C) 1991-2, RSA Data Security, Inc. Created 1991. All
rights reserved.

License to copy and use this software is granted provided that it
is identified as the "RSA Data Security, Inc. MD5 Message-Digest
Algorithm" in all material mentioning or referencing this software
or this function.

License is also granted to make and use derivative works provided
that such works are identified as "derived from the RSA Data
Security, Inc. MD5 Message-Digest Algorithm" in all material
mentioning or referencing the derived work.

RSA Data Security, Inc. makes no representations concerning either
the merchantability of this software or the suitability of this
software for any particular purpose. It is provided "as is"
without express or implied warranty of any kind.

These notices must be retained in any copies of any part of this
documentation and/or software.
 */


typedef unsigned int UINT4;

/* MD5 context. */
typedef struct {
  UINT4 state[4];                                   /* state (ABCD) */
  UINT4 count[2];        /* number of bits, modulo 2^64 (lsb first) */
  unsigned char buffer[64];                         /* input buffer */
} MD5_CTX;

/* MD5C.C - RSA Data Security, Inc., MD5 message-digest algorithm
 */

typedef void *POINTER;

/* Constants for MD5Transform routine.
 */
#define S11 7
#define S12 12
#define S13 17
#define S14 22
#define S21 5
#define S22 9
#define S23 14
#define S24 20
#define S31 4
#define S32 11
#define S33 16
#define S34 23
#define S41 6
#define S42 10
#define S43 15
#define S44 21

static void MD5Transform (UINT4 [4], unsigned char [64]);
static void Encode (unsigned char *, UINT4 *, unsigned int);
static void Decode (UINT4 *, unsigned char *, unsigned int);

static unsigned char PADDING[64] = {
  0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};

/* F, G, H and I are basic MD5 functions.
 */
#define F(x, y, z) (((x) & (y)) | ((~x) & (z)))
#define G(x, y, z) (((x) & (z)) | ((y) & (~z)))
#define H(x, y, z) ((x) ^ (y) ^ (z))
#define I(x, y, z) ((y) ^ ((x) | (~z)))

/* ROTATE_LEFT rotates x left n bits.
 */
#define ROTATE_LEFT(x, n) (((x) << (n)) | ((x) >> (32-(n))))

/* FF, GG, HH, and II transformations for rounds 1, 2, 3, and 4.
Rotation is separate from addition to prevent recomputation.
 */
#define FF(a, b, c, d, x, s, ac) { \
 (a) += F ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define GG(a, b, c, d, x, s, ac) { \
 (a) += G ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define HH(a, b, c, d, x, s, ac) { \
 (a) += H ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define II(a, b, c, d, x, s, ac) { \
 (a) += I ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }

/* MD5 initialization. Begins an MD5 operation, writing a new context.
 */
static void MD5Init (context)
MD5_CTX *context;                                        /* context */
{
  context->count[0] = context->count[1] = 0;
  /* Load magic initialization constants.
*/
  context->state[0] = 0x67452301;
  context->state[1] = 0xefcdab89;
  context->state[2] = 0x98badcfe;
  context->state[3] = 0x10325476;
}

/* MD5 block update operation. Continues an MD5 message-digest
  operation, processing another message block, and updating the
  context.
 */
static void MD5Update (context, input, inputLen)
MD5_CTX *context;                                        /* context */
unsigned char *input;                                /* input block */
unsigned int inputLen;                     /* length of input block */
{
  unsigned int i, index, partLen;

  /* Compute number of bytes mod 64 */
  index = (unsigned int)((context->count[0] >> 3) & 0x3F);

  /* Update number of bits */
  if ((context->count[0] += ((UINT4)inputLen << 3))
   < ((UINT4)inputLen << 3))
 context->count[1]++;
  context->count[1] += ((UINT4)inputLen >> 29);

  partLen = 64 - index;

  /* Transform as many times as possible.
*/
  if (inputLen >= partLen) {
 memcpy
   ((POINTER)&context->buffer[index], (POINTER)input, partLen);
 MD5Transform (context->state, context->buffer);

 for (i = partLen; i + 63 < inputLen; i += 64)
   MD5Transform (context->state, &input[i]);

 index = 0;
  }
  else
 i = 0;

  /* Buffer remaining input */
  memcpy
 ((POINTER)&context->buffer[index], (POINTER)&input[i],
  inputLen-i);
}

/* MD5 finalization. Ends an MD5 message-digest operation, writing the
  the message digest and zeroizing the context.
 */
static void MD5Final (digest, context)
unsigned char digest[16];                         /* message digest */
MD5_CTX *context;                                       /* context */
{
  unsigned char bits[8];
  unsigned int index, padLen;

  /* Save number of bits */
  Encode (bits, context->count, 8);

  /* Pad out to 56 mod 64.
*/
  index = (unsigned int)((context->count[0] >> 3) & 0x3f);
  padLen = (index < 56) ? (56 - index) : (120 - index);
  MD5Update (context, PADDING, padLen);

  /* Append length (before padding) */
  MD5Update (context, bits, 8);
  /* Store state in digest */
  Encode (digest, context->state, 16);

  /* Zeroize sensitive information.
*/
  memset ((POINTER)context, 0, sizeof (*context));
}

/* MD5 basic transformation. Transforms state based on block.
 */
static void MD5Transform (state, block)
UINT4 state[4];
unsigned char block[64];
{
  UINT4 a = state[0], b = state[1], c = state[2], d = state[3], x[16];

  Decode (x, block, 64);

  /* Round 1 */
  FF (a, b, c, d, x[ 0], S11, 0xd76aa478); /* 1 */
  FF (d, a, b, c, x[ 1], S12, 0xe8c7b756); /* 2 */
  FF (c, d, a, b, x[ 2], S13, 0x242070db); /* 3 */
  FF (b, c, d, a, x[ 3], S14, 0xc1bdceee); /* 4 */
  FF (a, b, c, d, x[ 4], S11, 0xf57c0faf); /* 5 */
  FF (d, a, b, c, x[ 5], S12, 0x4787c62a); /* 6 */
  FF (c, d, a, b, x[ 6], S13, 0xa8304613); /* 7 */
  FF (b, c, d, a, x[ 7], S14, 0xfd469501); /* 8 */
  FF (a, b, c, d, x[ 8], S11, 0x698098d8); /* 9 */
  FF (d, a, b, c, x[ 9], S12, 0x8b44f7af); /* 10 */
  FF (c, d, a, b, x[10], S13, 0xffff5bb1); /* 11 */
  FF (b, c, d, a, x[11], S14, 0x895cd7be); /* 12 */
  FF (a, b, c, d, x[12], S11, 0x6b901122); /* 13 */
  FF (d, a, b, c, x[13], S12, 0xfd987193); /* 14 */
  FF (c, d, a, b, x[14], S13, 0xa679438e); /* 15 */
  FF (b, c, d, a, x[15], S14, 0x49b40821); /* 16 */

 /* Round 2 */
  GG (a, b, c, d, x[ 1], S21, 0xf61e2562); /* 17 */
  GG (d, a, b, c, x[ 6], S22, 0xc040b340); /* 18 */
  GG (c, d, a, b, x[11], S23, 0x265e5a51); /* 19 */
  GG (b, c, d, a, x[ 0], S24, 0xe9b6c7aa); /* 20 */
  GG (a, b, c, d, x[ 5], S21, 0xd62f105d); /* 21 */
  GG (d, a, b, c, x[10], S22,  0x2441453); /* 22 */
  GG (c, d, a, b, x[15], S23, 0xd8a1e681); /* 23 */
  GG (b, c, d, a, x[ 4], S24, 0xe7d3fbc8); /* 24 */
  GG (a, b, c, d, x[ 9], S21, 0x21e1cde6); /* 25 */
  GG (d, a, b, c, x[14], S22, 0xc33707d6); /* 26 */
  GG (c, d, a, b, x[ 3], S23, 0xf4d50d87); /* 27 */
  GG (b, c, d, a, x[ 8], S24, 0x455a14ed); /* 28 */
  GG (a, b, c, d, x[13], S21, 0xa9e3e905); /* 29 */
  GG (d, a, b, c, x[ 2], S22, 0xfcefa3f8); /* 30 */
  GG (c, d, a, b, x[ 7], S23, 0x676f02d9); /* 31 */
  GG (b, c, d, a, x[12], S24, 0x8d2a4c8a); /* 32 */

  /* Round 3 */
  HH (a, b, c, d, x[ 5], S31, 0xfffa3942); /* 33 */
  HH (d, a, b, c, x[ 8], S32, 0x8771f681); /* 34 */
  HH (c, d, a, b, x[11], S33, 0x6d9d6122); /* 35 */
  HH (b, c, d, a, x[14], S34, 0xfde5380c); /* 36 */
  HH (a, b, c, d, x[ 1], S31, 0xa4beea44); /* 37 */
  HH (d, a, b, c, x[ 4], S32, 0x4bdecfa9); /* 38 */
  HH (c, d, a, b, x[ 7], S33, 0xf6bb4b60); /* 39 */
  HH (b, c, d, a, x[10], S34, 0xbebfbc70); /* 40 */
  HH (a, b, c, d, x[13], S31, 0x289b7ec6); /* 41 */
  HH (d, a, b, c, x[ 0], S32, 0xeaa127fa); /* 42 */
  HH (c, d, a, b, x[ 3], S33, 0xd4ef3085); /* 43 */
  HH (b, c, d, a, x[ 6], S34,  0x4881d05); /* 44 */
  HH (a, b, c, d, x[ 9], S31, 0xd9d4d039); /* 45 */
  HH (d, a, b, c, x[12], S32, 0xe6db99e5); /* 46 */
  HH (c, d, a, b, x[15], S33, 0x1fa27cf8); /* 47 */
  HH (b, c, d, a, x[ 2], S34, 0xc4ac5665); /* 48 */

  /* Round 4 */
  II (a, b, c, d, x[ 0], S41, 0xf4292244); /* 49 */
  II (d, a, b, c, x[ 7], S42, 0x432aff97); /* 50 */
  II (c, d, a, b, x[14], S43, 0xab9423a7); /* 51 */
  II (b, c, d, a, x[ 5], S44, 0xfc93a039); /* 52 */
  II (a, b, c, d, x[12], S41, 0x655b59c3); /* 53 */
  II (d, a, b, c, x[ 3], S42, 0x8f0ccc92); /* 54 */
  II (c, d, a, b, x[10], S43, 0xffeff47d); /* 55 */
  II (b, c, d, a, x[ 1], S44, 0x85845dd1); /* 56 */
  II (a, b, c, d, x[ 8], S41, 0x6fa87e4f); /* 57 */
  II (d, a, b, c, x[15], S42, 0xfe2ce6e0); /* 58 */
  II (c, d, a, b, x[ 6], S43, 0xa3014314); /* 59 */
  II (b, c, d, a, x[13], S44, 0x4e0811a1); /* 60 */
  II (a, b, c, d, x[ 4], S41, 0xf7537e82); /* 61 */
  II (d, a, b, c, x[11], S42, 0xbd3af235); /* 62 */
  II (c, d, a, b, x[ 2], S43, 0x2ad7d2bb); /* 63 */
  II (b, c, d, a, x[ 9], S44, 0xeb86d391); /* 64 */

  state[0] += a;
  state[1] += b;
  state[2] += c;
  state[3] += d;

  /* Zeroize sensitive information.
*/
  memset ((POINTER)x, 0, sizeof (x));
}

/* Encodes input (UINT4) into output (unsigned char). Assumes len is
  a multiple of 4.
 */
static void Encode (output, input, len)
unsigned char *output;
UINT4 *input;
unsigned int len;
{
  unsigned int i, j;

  for (i = 0, j = 0; j < len; i++, j += 4) {
 output[j] = (unsigned char)(input[i] & 0xff);
 output[j+1] = (unsigned char)((input[i] >> 8) & 0xff);
 output[j+2] = (unsigned char)((input[i] >> 16) & 0xff);
 output[j+3] = (unsigned char)((input[i] >> 24) & 0xff);
  }
}

/* Decodes input (unsigned char) into output (UINT4). Assumes len is
  a multiple of 4.
 */
static void Decode (output, input, len)
UINT4 *output;
unsigned char *input;
unsigned int len;
{
  unsigned int i, j;

  for (i = 0, j = 0; j < len; i++, j += 4)
 output[i] = ((UINT4)input[j]) | (((UINT4)input[j+1]) << 8) |
   (((UINT4)input[j+2]) << 16) | (((UINT4)input[j+3]) << 24);
}

/*************************************************************************/

/******** Our own high-level routines. ********/


#define XTOI(c) ((c)>9 ? (c)-'A'+10 : (c)-'0')

/* Encrypt `src' of length `len' and store the result in `dest'.  If the
 * resulting string would be longer than `size', return -1 and leave `dest'
 * unchanged; else return 0.
 */
int encrypt_md5(const char *src, int len, char *dest, int size)
{

    unsigned char buf[16];	
    MD5_CTX context;

    if (size < 17)
	return -1;

    memset(&context, 0, sizeof(context));
    memset(&buf, 0, sizeof(buf));
    
    MD5Init(&context);
    MD5Update(&context, src, len);
    MD5Final(buf, &context);
    memcpy(dest, buf, 16);
    return 0;
    
}

char* hex_str(unsigned char *str, int len)
{
  static char hexTab[] = "0123456789abcdef";
  static char buffer[200];
  int i;
  if(len>sizeof(buffer))
    abort();
  for(i=0;i<len;++i)
    {
      buffer[i*2]=hexTab[str[i] >> 4];
      buffer[(i*2)+1]=hexTab[str[i] & 0x0F];
    }
  return buffer;
} 

char* hex_bin(unsigned char *dest, char* source)
{
  int i=0,h1,h2;
  char *s = source;
  while(*s && *(s+1))
    {
      *s = toupper(*s);
      *(s+1) = toupper(*(s+1));
      if(*s>'9')
        h1 = 10 + (*s-'A');
      else 
        h1 = *s - '0'; 
      if(*(s+1)>'9')
        h2 = 10 + (*(s+1)-'A');
      else 
        h2 = *(s+1) - '0';         
      dest[i++] = (h1 << 4) + h2;
      s += 2;
    }
    dest[i]='\0';
    return dest;
} 
